app/controllers/bot_detection_controller.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+require 'http'
+
+# This controller has actions for issuing a challenge page for CloudFlare Turnstile product,
+# and then redirecting back to desired page.
+class BotDetectionController < ApplicationController
+  class_attribute :enabled, default: false # Must set to true to turn on at all
+
+  class_attribute :session_passed_good_for, default: 24.hours.ago
+
+  class_attribute :cf_turnstile_sitekey, default: '1x00000000000000000000AA' # a testing key that always passes
+  class_attribute :cf_turnstile_secret_key, default: '1x0000000000000000000000000000000AA' # a testing key always passes
+  class_attribute :cf_turnstile_js_url, default: 'https://challenges.cloudflare.com/turnstile/v0/api.js'
+  class_attribute :cf_turnstile_validation_url, default: 'https://challenges.cloudflare.com/turnstile/v0/siteverify'
+  class_attribute :cf_timeout, default: 3 # max timeout seconds waiting on Cloudfront Turnstile api
+  helper_method :cf_turnstile_js_url, :cf_turnstile_sitekey
+
+  # key stored in Rails session object with change passed confirmed
+  class_attribute :session_passed_key, default: 'bot_detection-passed'
+
+  # key in rack env that says challenge is required
+  class_attribute :env_challenge_trigger_key, default: 'bot_detect.should_challenge'
+
+  def self.bot_detection_enforce_filter(controller)
+    return unless enabled && !controller.session[session_passed_key].try { |date| Time.new(date) < session_passed_good_for }
+
+    return unless controller.request.get?
+
+    Rails.logger.info 'Redirecting for Turnstile'
+    controller.redirect_to '/challenge', status: 307
```

Comment:
Without the `dest` param there wouldn't be anywhere to redirect back to in the JS